src/c_api/c_api_symbolic.cc

```diff
@@ -810,6 +810,191 @@ int MXQuantizeSymbol(SymbolHandle sym_handle,
+      auto it_without_dtype = node_without_dtype_map.find(node_name);
+      if (it_model_params != model_params.end()) {
+        // need to update __dtype__ attribute if already set, else set it
+        if (it_with_dtype != node_name_dtype_map.end()) {
+          args[i]->attrs.dict[dtype_keyword] =
+              std::to_string(it_with_dtype->second);
+        } else {
+          CHECK(it_without_dtype != node_without_dtype_map.end())
+              << "make sure all nodes without dtype have properly been added "
+                 "in node_without_dtype_map";
+          args[i]->attrs.dict[dtype_keyword] =
+              std::to_string(it_without_dtype->second);
+        }
+      }
+    }
+  } else {
+    // if model params not provided, update __dtype__ for all inputs,
+    // which already had it set, don't touch the rest
+    for (size_t i = 0; i < args.size(); ++i) {
+      auto it = node_name_dtype_map.find(args[i]->attrs.name);
+      if (it != node_name_dtype_map.end()) {
+        if (args[i]->attrs.dict.find(dtype_keyword) !=
+            args[i]->attrs.dict.end()) {
+          args[i]->attrs.dict[dtype_keyword] = std::to_string(it->second);
+        }
+      }
+    }
+  }
+}
+
+int MXReducePrecisionSymbol(SymbolHandle sym_handle,
+                            SymbolHandle *ret_sym_handle,
+                            mx_uint num_args,
+                            const int *arg_type_data,
+                            mx_uint num_ind_ptr,
+                            const int* ind_ptr,
+                            const int *target_dtype,
+                            const mx_uint num_target_dtype_op_names,
+                            const mx_uint num_fp32_op_names,
+                            const mx_uint num_widest_dtype_op_names,
+                            const mx_uint num_conditional_fp32_op_names,
+                            const mx_uint num_excluded_symbols,
+                            const mx_uint num_model_params,
+                            const char **target_dtype_op_names,
+                            const char **fp32_op_names,
+                            const char **widest_dtype_op_names,
+                            const char **conditional_fp32_op_names,
+                            const char **excluded_symbols,
+                            const char **param_names,
+                            const char **param_vals,
+                            const char **model_param_names,
+                            const char **arg_names) {
+  nnvm::Symbol *result_sym = new nnvm::Symbol();
+  API_BEGIN();
+  nnvm::Symbol *sym = static_cast<nnvm::Symbol *>(sym_handle);
+  nnvm::Graph g = Symbol2Graph(*sym);
+  std::unordered_set<std::string> target_dtype_ops;
+  std::unordered_set<std::string> fp32_ops;
+  std::unordered_set<std::string> widest_dtype_ops;
+  std::unordered_set<std::string> excluded_syms;
+  std::unordered_set<std::string> model_params;
+  std::unordered_map<std::string,
```

Comment:
Suggest to add comment on what this thing represents, or maybe a typedef with a doc. It would help with code maintainability.